tests/samples_test.py

```diff
@@ -0,0 +1,420 @@
+)
+def test_nested_num_per_level(nested_samples, num_per_level):
+    """test number of points per level"""
+    assert numpy.array_equal(nested_samples.num_per_level, num_per_level)
+
+
+@pytest.mark.parametrize(
+    "nested_samples,start_level",
+    [
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 0), [0]),
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 1), [0, 1]),
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 2), [0, 1, 3]),
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 3), [0, 1, 3, 5]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 0), [0]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 1), [0, 1]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 2), [0, 1, 3]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 3), [0, 1, 3, 5]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 4), [0, 1, 3, 5, 9]),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            [0, 1, 3, 5, 9, 9],
+        ),
+        (smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 0), [0]),
+        (smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 1), [0, 1]),
+        (smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 2), [0, 1, 3]),
+    ],
+    ids=nested_sample_ids,
+)
+def test_nested_start_level(nested_samples, start_level):
+    """test number of points per level"""
+    assert numpy.array_equal(nested_samples.start_level, start_level)
+
+
+@pytest.mark.parametrize(
+    "nested_samples,end_level",
+    [
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 0), [1]),
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 1), [1, 3]),
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 2), [1, 3, 5]),
+        (smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 3), [1, 3, 5, 9]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 0), [1]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 1), [1, 3]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 2), [1, 3, 5]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 3), [1, 3, 5, 9]),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 4), [1, 3, 5, 9, 9]),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            [1, 3, 5, 9, 9, 17],
+        ),
+        (smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 0), [1]),
+        (smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 1), [1, 3]),
+        (smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 2), [1, 3, 9]),
+    ],
+    ids=nested_sample_ids,
+)
+def test_nested_end_level(nested_samples, end_level):
+    """test number of points per level"""
+    assert numpy.array_equal(nested_samples.end_level, end_level)
```

Comment:
You can combine these tests, it is fine to run multiple related asserts in a test

Reply:
I would ideally like to combine the `start_level`, `end_level`, and the `get_individual_levels` tests since the `get_individual_levels` test tests whether the values in the start_level and end_level get you the expected points for each level (basically testing if the `start_level` and `end_level` variables fulfill their purpose). However, the list of parameters in the `get_individual_levels` test isn't compatible with the `start_level` and `end_level` tests, so I'm not sure if there's a good way to do this.

Reply:
The parameters for `get_individual_levels` could be made the same as `start_level`, `end_level`, and `num_per_level`, but that would mean always checking the last level of the object. Alternatively, I could have the `num_per_level`, `start_level`, and `end_level`, tested for correctness, save the current `max_level` to a reference variable `check_level`, change the `max_level` to some constant, and then check the points at the level in `check_level`.

This isn't a test that check if `max_level` is set correctly, and in fact there is no such test for checking that changing `max_level` causes the points to update.

Reply:
Added tests for setting `max_level`, `degree`, and `frequency` to the initialization tests. I also implemented the strategy for merging the `start_level`, `end_level`, and `num_per_level` test with the test that checks that the values in `start_level` and `end_level` do correspond to the points expected in the individual levels.